src/dos/cdrom_ioctl_linux.cpp

```diff
@@ -113,4 +291,154 @@ bool CDROM_Interface_Ioctl::ReadSectorsHost([[maybe_unused]] void *buffer,
 	return false; /*TODO*/
 }
 
+void CDROM_Interface_Ioctl::InitAudio()
+{
+	// Find a a channel name that doesn't exist (channel names must be
+	// unique or they get overwritten)
+	mixer_channel_t exists;
+	int num                  = 0;
+	const std::string prefix = MixerChannelNamePrefix;
+	std::string full_name;
+	do {
+		full_name = prefix + std::to_string(num++);
+		exists    = MIXER_FindChannel(full_name.c_str());
+	} while (exists);
+
+	audio_buffer = std::make_unique<int16_t[]>(AudioBufferBytes / 2);
+	mixer_channel = MIXER_AddChannel(std::bind(&CDROM_Interface_Ioctl::CDAudioCallback,
+	                                           this,
+	                                           std::placeholders::_1),
+	                                 REDBOOK_PCM_FRAMES_PER_SECOND,
+	                                 full_name.c_str(),
+	                                 {ChannelFeature::Stereo,
+	                                  ChannelFeature::DigitalAudio});
+}
+
+void CDROM_Interface_Ioctl::CDAudioCallback(uint16_t requested_frames)
+{
+	if (audio_position == 0) {
+		if (sectors_remaining < 1) {
+			StopAudio();
+			return;
+		}
+		cdrom_read_audio cd = {};
+		cd.addr.lba         = current_sector;
+		cd.addr_format      = CDROM_LBA;
+		cd.buf = reinterpret_cast<uint8_t*>(audio_buffer.get());
+		// Maximum this ioctl can read in a single call is 1 second (75
+		// redbook frames)
+		cd.nframes = std::min(sectors_remaining, CD_FRAMES);
+		constexpr int PCMFramesPerRedbookFrame = BYTES_PER_RAW_REDBOOK_FRAME /
+		                                         BYTES_PER_REDBOOK_PCM_FRAME;
+		frames_in_buffer = cd.nframes * PCMFramesPerRedbookFrame;
+		current_sector += cd.nframes;
+		sectors_remaining -= cd.nframes;
+		if (cdrom_fd == -1 || ioctl(cdrom_fd, CDROMREADAUDIO, &cd) != 0) {
+			// Write out silence if something goes wrong
+			memset(audio_buffer.get(), 0, AudioBufferBytes);
+#ifdef DEBUG_IOCTL
+			LOG_WARNING("CDROM_IOCTL: CDAudioCallback: CDROMREADAUDIO ioctl failed");
+#endif
+		}
+	}
+
+	const size_t frames_available = frames_in_buffer - audio_position;
+	const size_t frames_to_write = std::min(static_cast<size_t>(requested_frames),
+	                                        frames_available);
+
+	// audio_position must be doubled since a frame is two int16_t's for stereo
+	// sound Does this need to be AddSamples_s16_nonnative on big endian?
+	mixer_channel->AddSamples_s16(frames_to_write,
```

Comment:
Ideally, we'd want a full float mixer path; see this ticket for further details:
https://github.com/dosbox-staging/dosbox-staging/issues/1720

But this is kind of a special case as you can only retrieve the raw CD Audio data as a stream of integers. Anyway, just something to keep in mind for the future.

Reply:
Nice one 🎉